api/README.md

```diff
@@ -0,0 +1,264 @@
+4. Existing clients need not be aware of your change in order for them to continue to function as they did previously, even when your change is in use.
+
+The main difference with the k8s API's definition of compatibility here is that it is not required to be able round-trip a change (convert to different API versions and back) with no loss of information, since contrary to k8s' API we don't need to store API objects here.
+
+As long as a change abides by the compatibility rules outlined above, one can simply amend the `proto` file defining that API version, and then regenerate all the Go boilerplate code.
+
+### How to add a new version to an existing API group
+
+Any change that cannot be done in a compatible way requires creating a new API version.
+
+Steps to add a new API version:
+1. define it its own new `api.proto` file
+2. generate the Go protobuf code
+3. update the API group's internal representations (in its `types.go` file) to be able to represent all of the group's version (the new and the old ones)
+3. add any needed conversion functions for all existing versions of this API group to account for the changes made at the previous step
+4. re-generate all of the Go boilerplate code
+5. now you can change the API group's server to add your new feature!
+
+### How to deprecate, and eventually remove...
+
+From the CSI [proxy KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md#csi-proxy-grpc-api-graduation-and-deprecation-policy):
+
+> In accordance with standard Kubernetes conventions, the above API will be introduced as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. Beyond a vN release in the future, new RPCs and enhancements to parameters will be introduced through vN+1alpha1 and graduate to vN+1beta1 and vN+1 stable versions as the new APIs mature.
+>
+> Members of CSIProxyService API may be deprecated and then removed from csi-proxy.exe in a manner similar to Kubernetes deprecation [policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) although maintainers will make an effort to ensure such deprecation is as rare as possible. After their announced deprecation, a member of CSIProxyService API must be supported:
+>
+> 1. 12 months or 3 releases (whichever is longer) if the API member is part of a Stable/vN version.
+> 2. 9 months or 3 releases (whichever is longer) if the API member is part of a Beta/vNbeta1 version.
+> 3. 0 releases if the API member is part of an Alpha/vNalpha1 version.
+> To continue running CSI Node Plugins that depend on an old version of csi-proxy.exe, vN, some of whose members have been removed, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
```

Comment:
Can you clarify here are we talking about the version of csi-proxy.exe, or the csi proxy API? My understanding here is that a single binary csi-proxy.exe will need to support multiple API versions, including deprecated ones until we reach the end of the deprecation period.

Reply:
Correct - in the vast majority of scenarios, we expect a single version of csi-proxy.exe to be running that supports the necessary API versions. We expect this to be the case for the first few releases. 

The above bit about multiple versions of csi-proxy.exe is geared towards a remote scenario in the future where an operator for some reason wants to keep running a really old CSI driver (that requires an API version that is no longer supported by the current version of csi-proxy.exe at that point in time) along with a modern CSI driver (that requires an API version that is only supported by current version of csi-proxy.exe at that point in time).

Reply:
Yeah I think distinguishing between the two would be good: what the supported case is vs unsupported but if you really want to this is what you could do

Reply:
Good point, clarifying.

Reply:
(also note for later: running 2 csi-proxy.exe at the same time will require small tweaks on named pipes paths)

Reply:
In the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md#csi-proxy-named-pipes) we proposed a probe and back-off mechanism to enable side-by-side execution (when necessary):

> For every version of the API supported, csi-proxy.exe will first probe for and then expose a \\.\pipe\... If during the initial 
probe phase, csi-proxy.exe determines that another process is already listening on a \\.\pipe\csi-proxy-v[N] named pipe, it will not try to create and listen on that named pipe. This allows multiple versions of csi-proxy.exe to run side-by-side if absolutely necessary to support multiple CSI Node Plugins that require widely different versions of CSIProxyService that no single version of csi-proxy.exe can support.

```

Reply:
Oh, missed that part. As of right now the server will crash if it can't create all the pipes it needs.

I'm not sure I love this though, because that has the potential to hide problems from admins. You want to know when your service can't listen on what it's supposed to be listening on.

How about having a flag for the server `--enabled-apis <group>-<version>,<group>-<version>,...` instead?

Reply:
Sounds good. More deterministic that way.

Reply:
It'll be an optional flag that you can pass to csi-proxy.exe. By default, all API groups/versions will be enabled.

Created https://github.com/kubernetes-csi/csi-proxy/issues/9